plugins/filter_aws/aws.c

```diff
@@ -182,191 +206,279 @@ static int cb_aws_init(struct flb_filter_instance *f_ins,
+    size_t i;
+    flb_sds_t tag_value = NULL;
+    size_t tag_value_len = 0;
+    size_t tag_value_path_len;
+    flb_sds_t tag_value_path;
+    flb_sds_t tmp;
+
+    /* initialize array for the tag values */
+    ctx->tag_values = flb_calloc(ctx->tags_count, sizeof(flb_sds_t));
+    if (!ctx->tag_values) {
+        flb_errno();
+        return -1;
+    }
+    ctx->tag_values_len = flb_calloc(ctx->tags_count, sizeof(size_t));
+    if (!ctx->tag_values_len) {
         flb_errno();
-        flb_http_client_destroy(client);
-        flb_upstream_conn_release(u_conn);
         return -1;
     }
 
-    *metadata = tmp;
-    *metadata_len = key == NULL ? client->resp.payload_size : strlen(tmp);
+    for (i = 0; i < ctx->tags_count; i++) {
+        /* fetch tag value using path: /latest/meta-data/tags/instance/{tag_name} */
+        tag_value_path_len = ctx->tag_keys_len[i] + 1 +
+                             strlen(FLB_AWS_IMDS_INSTANCE_TAG);
+        tag_value_path = flb_sds_create_size(tag_value_path_len);
+        if (!tag_value_path) {
+            flb_errno();
+            return -1;
+        }
+        tmp = flb_sds_printf(&tag_value_path, "%s/%s",
+                             FLB_AWS_IMDS_INSTANCE_TAG,
+                             ctx->tag_keys[i]);
+        if (!tmp) {
+            flb_errno();
+            flb_sds_destroy(tag_value_path);
+            return -1;
+        }
+        tag_value_path = tmp;
+
+        ret = flb_aws_imds_request(ctx->client_imds, tag_value_path, &tag_value,
+                                   &tag_value_len);
+        if (ret < 0) {
+            flb_sds_destroy(tag_value_path);
+            if (ret == -2) {
+                flb_plg_error(ctx->ins, "no value for tag %s", ctx->tag_keys[i]);
+            } else {
+                flb_plg_error(ctx->ins, "could not fetch value for tag %s",
+                              ctx->tag_keys[i]);
+            }
+            return ret;
+        }
 
-    flb_http_client_destroy(client);
-    flb_upstream_conn_release(u_conn);
-    return 0;
-}
+        ctx->tag_values[i] = tag_value;
+        ctx->tag_values_len[i] = tag_value_len;
 
-static int get_metadata(struct flb_filter_aws *ctx, char *metadata_path,
-                        flb_sds_t *metadata, size_t *metadata_len)
-{
-    return get_metadata_by_key(ctx, metadata_path, metadata,
-                               metadata_len, NULL);
+        flb_sds_destroy(tag_value_path);
+    }
+
+    return 0;
 }
 
-/* get VPC metadata, it called IMDS twice.
- * First is for getting the Mac ID and combine into the path for VPC.
- * Second call is using the VPC path to get the VPC id
- */
-static int get_vpc_metadata(struct flb_filter_aws *ctx)
+static int get_ec2_tags(struct flb_filter_aws *ctx)
 {
     int ret;
-    flb_sds_t mac_id = NULL;
-    size_t len = 0;
 
-    /* get EC2 instance Mac id first before getting VPC id */
-    ret = get_metadata(ctx, FLB_FILTER_AWS_IMDS_MAC_PATH, &mac_id, &len);
+    ctx->tags_fetched = false;
```

Comment:
In Fluent Bit, we use `int` for boolean and use the values `FLB_TRUE` and `FLB_FALSE`.